dialogue-core/src/main/java/com/palantir/dialogue/core/FixedLimitedChannel.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.palantir.dialogue.core;
+
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.palantir.dialogue.Endpoint;
+import com.palantir.dialogue.Request;
+import com.palantir.dialogue.Response;
+import com.palantir.logsafe.SafeArg;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Channel with a fixed number of available permits. This can be used to enforce a per-route request limit.
+ */
+final class FixedLimitedChannel implements LimitedChannel {
+    private static final Logger log = LoggerFactory.getLogger(FixedLimitedChannel.class);
+
+    private final LimitedChannel delegate;
+    private final AtomicInteger availablePermits;
```

Comment:
Somehow my brain prefers the inversion of this, where this number starts at zero and counts the number of active requests?

Reply:
Just a thought - do you think a LongAdder would reduce contention and increase throughput here? I'm just thinking if all the witchcraft worker threads in a server are constantly blasting though these FixedLimitedChannels and there are two atomicinteger writes per call to maybeExecute, perhaps the more expensive `LongAdder#sum()` call is worth it?

Quite possibly over-optimizing, just wanted to throw the idea out there

Reply:
LongAdder doesn't provide any concurrency guarantees, I believe reading the value is more expensive though writing is cheap. It works well for things like counters/meters because they have orders of magnitude more writes than reads, and the value doesn't need to be precise.

There's always AtomicIntegerFieldUpdater, but it performs similarly to AtomicInteger with lower allocation overhead for creation, we could use it if we wanted to optimize around fitting the entire channel chain into a cache line... but that's likely a bit too far.